tool/tctl/common/top/command.go

```diff
@@ -40,27 +42,68 @@ type Command struct {
 	refreshPeriod time.Duration
 }
 
+const defaultDiagAddr = "http://127.0.0.1:3000"
+
 // Initialize sets up the "tctl top" command.
 func (c *Command) Initialize(app *kingpin.Application, _ *tctlcfg.GlobalCLIFlags, config *servicecfg.Config) {
 	c.config = config
 	c.top = app.Command("top", "Report diagnostic information.")
-	c.top.Arg("diag-addr", "Diagnostic HTTP URL").Default("http://127.0.0.1:3000").StringVar(&c.diagURL)
+	c.top.Arg("diag-addr", "Diagnostic HTTP URL").StringVar(&c.diagURL)
 	c.top.Arg("refresh", "Refresh period").Default("5s").DurationVar(&c.refreshPeriod)
 }
 
+type MetricsClient interface {
+	GetMetrics(ctx context.Context) (map[string]*dto.MetricFamily, error)
+}
+
+func (c *Command) newMetricsClient(ctx context.Context) (string, MetricsClient, error) {
+	if c.diagURL != "" {
+		clt, err := diag.NewClient(c.diagURL)
+		return c.diagURL, clt, trace.Wrap(err)
+	}
+
+	// Try the local UNIX debug service client first.
+	debugClient := debug.NewClient(c.config.DataDir)
+	_, debugErr := debugClient.GetMetrics(ctx)
+	if debugErr == nil {
+		return debugClient.SocketPath(), debugClient, nil
+	}
+	debugErr = trace.Wrap(debugErr, "retrieving metrics from debug service")
+
+	// Try default diagnostic address
+	diagClient, defErr := diag.NewClient(defaultDiagAddr)
+	if defErr != nil {
+		return "", nil, trace.NewAggregate(
+			trace.Errorf("unable to connect to Teleport service"),
+			trace.Wrap(defErr, "creating diagnostics client for default address %q", defaultDiagAddr),
+			debugErr)
+	}
+
+	_, defErr = diagClient.GetMetrics(ctx)
+	if defErr == nil {
+		return defaultDiagAddr, diagClient, nil
+	}
+
+	return "", nil, trace.NewAggregate(
+		trace.Errorf("unable to connect to Teleport service"),
+		trace.Wrap(defErr, "getting metrics from diagnostics client at default address %q", defaultDiagAddr),
+		debugErr,
+	)
```

Comment:
How does this get rendered to the user? In terms of error semantics this should be something like the latter, right?

```suggestion
	return "", nil, trace.Wrap(trace.NewAggregate(
		trace.Wrap(defErr, "getting metrics from diagnostics client at default address %q", defaultDiagAddr),
		debugErr,
	), "connecting to Teleport service")
```

Reply:
Agreed, for posterity this is the current behaviour:
```
ERROR: unable to connect to Teleport service, getting metrics from diagnostics client at default address "http://127.0.0.1:3000"
        Get "http://127.0.0.1:3000/metrics": dial tcp 127.0.0.1:3000: connect: connection refused, retrieving metrics from debug service
        Get "http://debug/metrics": dial unix /var/lib/teleport/debug.sock: connect: connection refused
```
and this is with your suggested change:

```
ERROR: connecting to Teleport service
        getting metrics from diagnostics client at default address "http://127.0.0.1:3000"
        Get "http://127.0.0.1:3000/metrics": dial tcp 127.0.0.1:3000: connect: connection refused, retrieving metrics from debug service
        Get "http://debug/metrics": dial unix /var/lib/teleport/debug.sock: connect: connection refused
```
It's a bit clearer to identify the top level error this way.